pkg/conversion/mapi2capi/machineset.go

```diff
@@ -55,9 +58,250 @@ func fromMAPIMachineSetToCAPIMachineSet(mapiMachineSet *mapiv1.MachineSet) (*clu
 			// MachineNamingStrategy: // Not supported in MAPI, remains nil. No equivalent field in MAPI MachineSet.
 			// AuthoritativeAPI: // Ignore, this is part of the conversion mechanism.
 		},
+		Status: convertMAPIMachineSetToCAPIMachineSetStatus(mapiMachineSet),
 	}
 
 	errs = append(errs, handleUnsupportedMAPIObjectMetaFields(field.NewPath("spec", "template", "metadata"), mapiMachineSet.Spec.Template.ObjectMeta)...)
 
 	return capiMachineSet, errs.ToAggregate()
 }
+
+// convertMAPIMachineSetToCAPIMachineSetStatus converts a MAPI MachineSet to CAPI MachineSetStatus.
+func convertMAPIMachineSetToCAPIMachineSetStatus(mapiMachineSet *mapiv1.MachineSet) clusterv1.MachineSetStatus {
+	capiStatus := clusterv1.MachineSetStatus{
+		Replicas:             mapiMachineSet.Status.Replicas,
+		FullyLabeledReplicas: mapiMachineSet.Status.FullyLabeledReplicas,
+		ReadyReplicas:        mapiMachineSet.Status.ReadyReplicas,
+		AvailableReplicas:    mapiMachineSet.Status.AvailableReplicas,
+		ObservedGeneration:   mapiMachineSet.Generation, // Set the observed generation to the current CAPI MachineSet generation.
+		Conditions:           convertMAPIMachineSetConditionsToCAPIMachineSetConditions(mapiMachineSet),
+		V1Beta2:              convertMAPIMachineSetStatusToCAPIMachineSetV1Beta2Status(mapiMachineSet),
+	}
+
+	// Convert ErrorReason/ErrorMessage to FailureReason/FailureMessage
+	if mapiMachineSet.Status.ErrorReason != nil {
+		capiStatus.FailureReason = convertMAPIErrorReasonToCAPIFailureReason(*mapiMachineSet.Status.ErrorReason)
+	}
+
+	if mapiMachineSet.Status.ErrorMessage != nil {
+		capiStatus.FailureMessage = mapiMachineSet.Status.ErrorMessage
+	}
+
+	// unused fields from MAPI MachineSetStatus
+	// - Selector: label selection is different between CAPI and MAPI.
+	// - AuthoritativeAPI: this is part of the conversion mechanism, it is not used in CAPI.
+	// - SynchronizedGeneration: this is part of the conversion mechanism, it is not used in CAPI.
+
+	return capiStatus
+}
+
+func convertMAPIMachineSetStatusToCAPIMachineSetV1Beta2Status(mapiMachineSet *mapiv1.MachineSet) *clusterv1.MachineSetV1Beta2Status {
+	return &clusterv1.MachineSetV1Beta2Status{
+		ReadyReplicas:     ptr.To(mapiMachineSet.Status.ReadyReplicas),
+		AvailableReplicas: ptr.To(mapiMachineSet.Status.AvailableReplicas),
+		UpToDateReplicas:  ptr.To(mapiMachineSet.Status.FullyLabeledReplicas), // Should be ok to do this.
+		Conditions:        convertMAPIMachineSetConditionsToCAPIMachineSetV1Beta2StatusConditions(mapiMachineSet),
```

Comment:
Optional: CAPI patches conditions and ensures a specific order of the conditions (for easier readability top-down for a user).

We could consider doing this too, or simply ignore.